src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs

```diff
@@ -1,77 +1,114 @@
-using System.Collections.Concurrent;
 using Microsoft.AspNetCore.Html;
 using OrchardCore.DisplayManagement.Implementation;
 using OrchardCore.Environment.Extensions.Features;
 
 namespace OrchardCore.DisplayManagement.Descriptors;
 
-public class FeatureShapeDescriptor : ShapeDescriptor
+public sealed class FeatureShapeDescriptor : ShapeDescriptor
 {
     public FeatureShapeDescriptor(IFeatureInfo feature, string shapeType)
     {
         Feature = feature;
         ShapeType = shapeType;
     }
 
-    public IFeatureInfo Feature { get; private set; }
+    public IFeatureInfo Feature { get; }
 }
 
-public class ShapeDescriptorIndex : ShapeDescriptor
+public sealed class ShapeDescriptorIndex : ShapeDescriptor
 {
-    private readonly List<FeatureShapeDescriptor> _alternationDescriptors = [];
-    private readonly List<string> _wrappers = [];
-    private readonly List<string> _bindingSources = [];
-    private readonly Dictionary<string, ShapeBinding> _bindings = new(StringComparer.OrdinalIgnoreCase);
-    private readonly List<Func<ShapeCreatingContext, Task>> _creatingAsync = [];
-    private readonly List<Func<ShapeCreatedContext, Task>> _createdAsync = [];
-    private readonly List<Func<ShapeDisplayContext, Task>> _displayingAsync = [];
-    private readonly List<Func<ShapeDisplayContext, Task>> _processingAsync = [];
-    private readonly List<Func<ShapeDisplayContext, Task>> _displayedAsync = [];
+    private readonly List<FeatureShapeDescriptor> _alternationDescriptors;
```

Comment:
I am pretty sure `alterNation` is a type, and I meant alteration. see https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/ShapeAlteration.cs#L5